### PR TITLE
Created the Match Alpha node

### PR DESCRIPTION
**Pull request by AlexBxl**
# Description

* This node finds the alpha value that can be used to composite two colors to match a reference third color.
* If the foreground and background hues are different, it returns NaN.
* If the resulting alpha is outside the valid range (0-1), it returns NaN.
* If the background and reference are the same, it returns 0.
* Otherwise it returns the found value with number of decimals from the precision input.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: packages/graph-engine/src/nodes/color/matchAlpha.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+export default class NodeDefinition extends Node {
+	static title = 'Match Alpha';
+	static type = 'studio.tokens.color.matchAlpha';
+	static description =
+		'Finds the alpha value that, when used to blend the foreground and background colors, will result in the reference color.';
+	constructor(props: INodeDefinition) {
+		super(props);
+
+		this.addInput('foreground', {
+			type: {
```
Comment (SorsOps) [inline packages/graph-engine/src/nodes/color/matchAlpha.ts:15]: If you're not altering the type , you can just do 
`type: ColorSchema`

File: packages/graph-engine/src/nodes/color/matchAlpha.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+		) {
+			alpha = 0;
+		} else {
+			// find the matching alpha for each channel
+			let ar = (ref.r - bg.r) / (fg.r - bg.r);
+			let ag = (ref.g - bg.g) / (fg.g - bg.g);
+			let ab = (ref.b - bg.b) / (fg.b - bg.b);
+
+			// set out-of-range alphas to NaN
+			if (ar < 0 || ar > 1) ar = Number.NaN;
```
Comment (SorsOps) [inline packages/graph-engine/src/nodes/color/matchAlpha.ts:67]: You can make this a single function
function normalize(value){
 if (value>=0 && ar<=1){
 return value
 }
 return Number.Nan;
} 

and then 

`const ar =normalize((ref.r - bg.r) / (fg.r - bg.r));`

File: packages/graph-engine/src/nodes/color/matchAlpha.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+				if (Math.abs(ar - ab) < precision) {
+					alpha = (ar + ab) / 2;
+				}
+			} else {
+				alpha = ar || ag || ab;
+			}
+		}
+
+		// round the result to match the precision
+		const decimals = Math.max(0, -Math.floor(Math.log10(precision)));
```
Comment (SorsOps) [inline packages/graph-engine/src/nodes/color/matchAlpha.ts:98]: I think we have util that handles this already?

File: packages/graph-engine/tests/suites/nodes/color/matchAlpha.test.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
+import { Graph } from '../../../../src/graph/graph.js';
+import { describe, expect, test } from 'vitest';
+import { getAllOutputs } from '../../../../src/utils/node.js';
+import { isNaN } from 'lodash-es';
```
Comment (SorsOps) [inline packages/graph-engine/tests/suites/nodes/color/matchAlpha.test.ts:4]: There is a native isNaN we can use

Comment (AlexBxl) [reply packages/graph-engine/tests/suites/nodes/color/matchAlpha.test.ts:4]: That's what I did first but it kept getting the red squiggles, until I used autocomplete and it found this one, then it worked.

Comment (SorsOps) [reply packages/graph-engine/tests/suites/nodes/color/matchAlpha.test.ts:4]: Did you use `Number.isNaN` ?

**Comment by AlexBxl** [comment]
I updated how the threshold check is done, the new algorithm is as follows:
image
I find the expected composited result using the found alpha value. If the composite is too far from the reference, the result is not valid.
Currently I only added one precision input. In the future, based on actual usage, we can add a separate testing threshold.

File: packages/graph-engine/src/nodes/color/matchAlpha.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import { Node } from '../../programmatic/node.js';
+import { setToPrecision } from '@/utils/precision.js';
+import Color from 'colorjs.io';
+
+export default class NodeDefinition extends Node {
+	static title = 'Match Alpha';
+	static type = 'studio.tokens.color.matchAlpha';
+	static description =
+		'Finds the alpha value that, when used to blend the foreground and background colors, will result in the reference color.';
+	constructor(props: INodeDefinition) {
```
Comment (SorsOps) [inline packages/graph-engine/src/nodes/color/matchAlpha.ts:17]: You can use the declare syntax to make this more obvious for typescript users
eg
```
	declare inputs: ToInput<{
		top: number;
		right: number;
		bottom: number;
		left: number;
	}>;
	declare outputs: ToOutput<{
		value: string;
	}>;
	```

Comment (AlexBxl) [reply packages/graph-engine/src/nodes/color/matchAlpha.ts:17]: How to put default values in this format?

**Pull request merged**